### PR TITLE
Join flow: custom error messages, retry rules for various particular registration error causes

### PR DESCRIPTION
**Pull request by benjiwheeler**
### Resolves:

Step towards resolving https://github.com/LLK/scratch-www/issues/3053

### Changes:

* Added error strings for particular join flow errors
* improve css formatting of welcome, error step text
* added validation functions for server response error code validation
* refactor error handling in join flow's handleRegistrationResponse function, to split logic into multiple smaller functions
* adds validation function to provide specific l10n response strings for a handful of particular server errors
* reset registrationError content when registration is retried
* revise, add tests for:
   * join flow component
   * validation library
   * join flow registration error component

### Test Coverage:

* added tests for join-flow.jsx and for validate.js

image

### CODE REVIEW

File: src/components/join-flow/join-flow.jsx [hunk 1]
```diff
@@ -42,15 +43,17 @@ class JoinFlow extends React.Component {
         this.state = this.initialState;
     }
     canTryAgain () {
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: maybe distinguish the two canTryAgains, by name?

File: src/components/join-flow/registration-error-step.jsx [hunk 2]
```diff
@@ -24,24 +25,42 @@ class RegistrationErrorStep extends React.Component {
...
 RegistrationErrorStep.propTypes = {
-    canTryAgain: PropTypes.bool,
+    canTryAgain: PropTypes.bool.isRequired,
     errorMsg: PropTypes.string,
     intl: intlShape,
-    onSubmit: PropTypes.func
+    onSubmit: PropTypes.func.isRequired
 };
 
 const IntlRegistrationErrorStep = injectIntl(RegistrationErrorStep);
```
Comment (benjiwheeler) [inline src/components/join-flow/registration-error-step.jsx]: Maybe add default value for canTryAgain, instead of isRequired?

File: src/components/join-flow/join-flow.jsx [hunk 3]
```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
...
+            vals.forEach(val => {
+                if (customErrMsg.length) customErrMsg += '; ';
+                customErrMsg += `${key}: ${val}`;
+            });
+        });
+        if (!customErrMsg) return null; // if no key-val pairs
+        const problemsStr = this.props.intl.formatMessage({id: 'registration.problemsAre'});
+        return `${problemsStr} "${customErrMsg}"`;
+    }
+    getRegistrationSuccess (err, body, res) {
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: Rename something that indicates this returns a bool. is/wasRegistrationSuccessfull, wasUserCreated, or something like that.

File: src/components/join-flow/join-flow.jsx [hunk 4]
```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
...
-                    if (!errStr.length && body[0].msg) errStr = body[0].msg;
+            const success = this.getRegistrationSuccess(err, body, res);
+            if (success) {
+                this.props.refreshSession();
+                this.setState({step: this.state.step + 1});
+                return;
+            }
+            // now we know there was some error.
+            // if the error was client-side, prompt user to try again
+            if (err || (res.statusCode >= 400 && res.statusCode < 500)) {
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: Maybe change this to retry on 500s too?

File: src/components/join-flow/join-flow.jsx [hunk 5]
```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
             this.handleSubmitRegistration(this.state.formData);
         });
     }
+    getBodyErrors (err, body, res) {
+        return (!err && res.statusCode === 200 && body && body[0] && body[0].errors);
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: what if this iterated through and produced array of errors like

```
[
   {fieldName: 'username', errorStr: 'invalid username'},
   {fieldName: 'username', errorStr: 'username too short'},
   {fieldName: 'captcha', errorStr: 'invalid captcha token'}
]
```

File: src/components/join-flow/join-flow.jsx [hunk 6]
```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
...
+            // if the error was client-side, prompt user to try again
+            if (err || (res.statusCode >= 400 && res.statusCode < 500)) {
+                this.setState({registrationError: {canTryAgain: true}});
+                return;
+            }
+            // now we know error was server-side.
+            // if server provided us info on why registration failed,
+            // build a summary explanation string
+            let errorMsg = null;
+            const bodyErrors = this.getBodyErrors(err, body, res);
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: if getBodyErrors instead returns more understandable object, then we can just say if (length === 1) look up single error str...

File: src/components/join-flow/join-flow.jsx [hunk 7]
```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
...
+    getSingleError (bodyErrors) {
+        if (Object.keys(bodyErrors).length === 1) {
+            const fieldName = Object.keys(bodyErrors)[0];
+            if (bodyErrors[fieldName].length === 1) {
+                return {fieldName: fieldName, errorStr: bodyErrors[fieldName][0]};
+            }
+        }
+        return null;
+    }
+    getCustomErrMsg (bodyErrors) {
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: expect differently formed object

File: src/components/join-flow/registration-error-step.jsx [hunk 8]
```diff
@@ -24,24 +25,42 @@ class RegistrationErrorStep extends React.Component {
...
                     this.props.intl.formatMessage({id: 'general.tryAgain'}) :
                     this.props.intl.formatMessage({id: 'general.startOver'})
                 }
-                title={this.props.intl.formatMessage({id: 'registration.generalError'})}
+                title={this.props.intl.formatMessage({id: 'general.error'})}
+                titleClassName="join-flow-error-title"
                 onSubmit={this.handleSubmit}
-            />
+            >
+                <p className="join-flow-instructions">
```
Comment (benjiwheeler) [inline src/components/join-flow/registration-error-step.jsx]: check what <p> is doing... prolly use <div> instead?

File: test/unit/components/join-flow.test.jsx [hunk 9]
```diff
@@ -244,4 +303,139 @@ describe('JoinFlow', () => {
...
+            refreshSession: jest.fn()
+        };
+        const joinFlowInstance = getJoinFlowWrapper(props).instance();
+        joinFlowInstance.handleRegistrationResponse(null, responseBodySuccess, {statusCode: 200});
+        expect(joinFlowInstance.state.registrationError).toEqual(null);
+        expect(joinFlowInstance.state.step).toEqual(1);
+        expect(joinFlowInstance.state.waiting).toBeFalsy();
+    });
+
+    test('handleRegistrationResponse when passed body with preset server error', () => {
```
Comment (benjiwheeler) [inline test/unit/components/join-flow.test.jsx]: move these tests up to other related tests

File: test/unit/lib/validate.test.js [hunk 10]
```diff
@@ -111,4 +111,11 @@ describe('unit test lib/validate.js', () => {
         response = validate.validateEmailLocally('much."more unusual"@example.com');
         expect(response).toEqual({valid: false, errMsgId: 'registration.validationEmailInvalid'});
     });
+
+    test('get responseErrorMsg', () => {
```
Comment (benjiwheeler) [inline test/unit/lib/validate.test.js]: add test for string we don't cover; response should be null

File: src/components/join-flow/join-flow.jsx (hunk 6 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done

File: src/components/join-flow/join-flow.jsx (hunk 3 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done; called it `registrationIsSuccessful`

File: src/components/join-flow/join-flow.jsx (hunk 7 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done

File: src/components/join-flow/join-flow.jsx (hunk 1 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done; called the property `errorAllowsTryAgain`

File: src/components/join-flow/join-flow.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done

File: src/components/join-flow/join-flow.jsx (hunk 4 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Done

File: src/components/join-flow/registration-error-step.jsx (hunk 8 above)
Comment (benjiwheeler) [reply src/components/join-flow/registration-error-step.jsx]: Done

File: src/components/join-flow/registration-error-step.jsx (hunk 2 above)
Comment (benjiwheeler) [reply src/components/join-flow/registration-error-step.jsx]: Done

File: test/unit/components/join-flow.test.jsx [hunk 11]
```diff
@@ -244,4 +303,138 @@ describe('JoinFlow', () => {
...
+
+    test('handleRegistrationResponse when passed status 500', () => {
+        const joinFlowInstance = getJoinFlowWrapper().instance();
+        joinFlowInstance.handleRegistrationResponse(null, responseBodyMultipleErrs, {statusCode: 500});
+        expect(joinFlowInstance.state.registrationError).toEqual({
+            errorAllowsTryAgain: true
+        });
+    });
+
+    test('handleErrorNext', () => {
```
Comment (picklesrus) [inline test/unit/components/join-flow.test.jsx]: This test looks forgotten :)

File: src/components/join-flow/join-flow.jsx [hunk 12]
```diff
@@ -61,59 +64,113 @@ class JoinFlow extends React.Component {
...
+                errorMsg = this.getCustomErrMsg(errorsFromResponse);
+            }
             this.setState({
-                registrationError: errStr ||
-                    `${this.props.intl.formatMessage({
-                        id: 'registration.generalError'
-                    })} (${res.statusCode})`
+                registrationError: {
+                    errorAllowsTryAgain: false,
+                    errorMsg: errorMsg
```
Comment (picklesrus) [inline src/components/join-flow/join-flow.jsx]: What happens when errorMsg is null? As far as I can tell this eventually gets sent to the RegistrationError component but it doesn't check whether the message exists before it tries to do something with it.  I may have just lost the trace of it though.  It might also be worth adding a test case for that if there isn't one already.

File: src/components/join-flow/join-flow.jsx [hunk 13]
```diff
@@ -42,15 +43,17 @@ class JoinFlow extends React.Component {
         this.state = this.initialState;
     }
     canTryAgain () {
-        return (this.state.numAttempts <= 1);
+        return (this.state.registrationError.canTryAgain && this.state.numAttempts <= 1);
```
[the remark this answers is not in the document]
Comment (picklesrus) [reply src/components/join-flow/join-flow.jsx]: whoops ignore this - it is left over from when I originally reviewing last week.

File: test/unit/lib/validate.test.js (hunk 10 above)
Comment (benjiwheeler) [reply test/unit/lib/validate.test.js]: Done

File: test/unit/components/join-flow.test.jsx (hunk 11 above)
Comment (benjiwheeler) [reply test/unit/components/join-flow.test.jsx]: Ooops! Removed!

File: test/unit/components/join-flow.test.jsx (hunk 9 above)
Comment (benjiwheeler) [reply test/unit/components/join-flow.test.jsx]: Done

File: src/components/join-flow/join-flow.jsx (hunk 12 above)
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: registrationErrorStep's render() function has the line

```
                {this.props.errorMsg && (
```
...which tests for errorMsg's truthiness.

This approach is simpler than conditionally including or excluding the prop when we set the registrationError object.

**Comment by picklesrus** [comment]
LGTM once you resolve conflicts.
